#! /bin/sh
patch -p1 -l -f $* < $0
exit $?

From 45eb36a00e0bb39fd02cb065e79cbc53945e7103 Mon Sep 17 00:00:00 2001
From: Werner Koch <wk@gnupg.org>
Date: Mon, 19 Aug 2013 20:29:25 +0200
Subject: [PATCH] Fix wrong use of GPGME_EVENT_DONE.

* src/gpacontext.c (gpa_context_event_cb): Fix use of TYPE_DATA.  Add
debug output.
--

With GPGME commit c8e934b2 (2009-10-26) gpgme_io_event_done_data_t was
introduced to replace the use of gpg_error_t with GPGME_EVENT_DONE.
Unfortunately this was not documented.  Maybe at that time the event
code was considered internal and its use in GPA was not known.  Too
bad.
---
 src/gpacontext.c |   13 +++++++++----
 1 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/src/gpacontext.c b/src/gpacontext.c
index e021d5f..b049f84 100644
--- a/src/gpacontext.c
+++ b/src/gpacontext.c
@@ -405,16 +405,21 @@ static void
 gpa_context_event_cb (void *data, gpgme_event_io_t type, void *type_data)
 {
   GpaContext *context = data;
-  gpg_error_t *err;
-
+  gpg_error_t err, op_err;
+
   switch (type)
     {
     case GPGME_EVENT_START:
       g_signal_emit (context, signals[START], 0);
       break;
     case GPGME_EVENT_DONE:
-      err = type_data;
-      g_signal_emit (context, signals[DONE], 0, *err);
+      err = ((gpgme_io_event_done_data_t)type_data)->err;
+      op_err = ((gpgme_io_event_done_data_t)type_data)->op_err;
+      g_debug ("EVENT_DONE: err=%s op_err=%s",
+               gpg_strerror (err), gpg_strerror (op_err));
+      if (!err)
+        err = op_err;
+      g_signal_emit (context, signals[DONE], 0, err);
       break;
     case GPGME_EVENT_NEXT_KEY:
       g_signal_emit (context, signals[NEXT_KEY], 0, type_data);
--
1.7.7.1
